### PR TITLE
[WIP] drivers: clock_control: initial support for GD32

### LINKED ISSUE #49074: GD32: Use clocks instead rcu-periph-clock property
**Is your enhancement proposal related to a problem? Please describe.**
The `rcu-periph-clock` property is not well align with devicetree and should be replaced. The correct property for that use is `clocks`.
This is a step closer to have a full clock capability on the GD32 SoCs family. 

**Describe the solution you'd like**
Replace `rcu-periph-clock` property by correct `clocks` property. The change itself not require a implementation of clock driver to allow the fix.

**Additional context**
The solution can mimics the following [implementation](https://elixir.bootlin.com/linux/v5.19/source/Documentation/devicetree/bindings/spi/atmel,quadspi.yaml#L85) from Linux. A reference from `&rcu` node + `RCU_TYPE_PERIPHERAL` + clock identificator. The change to correct the property not seems to represent a high effort in the first look.

CC: @gmarull @cameled @soburi

### PR DESCRIPTION
**Pull request by gmarull**
This PR (WIP) adds a new clock control driver for GD32 SoCs. It has been briefly tested on GD32F450I-EVAL board for now.

TODO:

- [ ] Port remaining drivers (uart, pwm, dac...)
- [ ] Generalize driver to support more series (VF103, F403, F350, etc.)
- [ ] Add tests

Fixes #49074

cc: @cameled @soburi

### CODE REVIEW

File: drivers/clock_control/clock_control_gd32.c [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	/* enable PLL and wait until stable */
+	RCU_CTL |= RCU_CTL_PLLEN;
+	while((RCU_CTL & RCU_CTL_PLLSTB) == 0U)
+		;
+}
+#endif /* GD32_PLL_IS_ENABLED */
+
+#ifdef GD32_PMU_ENABLE_HIGH_DRIVER
```
Comment (soburi) [inline drivers/clock_control/clock_control_gd32.c:152]: maybe #if, not #ifdef

File: drivers/clock_control/clock_control_gd32.c [hunk 2]
```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (c) 2022 Teslabs Engineering S.L.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <devicetree.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/gd32.h>
+#include <dt-bindings/clock/gd32_rcu_f4xx.h>
```
Comment (soburi) [inline drivers/clock_control/clock_control_gd32.c:10]: need to consider smart way to switch include files by SoC models

**Comment by soburi** [comment]
@gmarull 
GD32VF103 support patch post to https://github.com/teslabs/zephyr/pull/1.

File: drivers/clock_control/clock_control_gd32.c [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+	SystemCoreClock >>= ahb_exp[GD32_AHB_PRESCALER];
+}
+
+static int gd32_clock_control_on(const struct device *dev,
+				 clock_control_subsys_t sys)
+{
+	uint16_t cfg = *(uint16_t *)sys;
+
+	GD32_RCU_CFG_REG(cfg) |= BIT(GD32_RCU_CFG_BIT(cfg));
+
```
Comment (soburi) [inline drivers/clock_control/clock_control_gd32.c:202]: I think it is a need to wait for clock stable here. 
(Or we can change this function as async_on to make compliant the API spec, but we have no method to know stabilized the clocks asynchronously... :-( )

**Comment by cameled** [comment]
Collected some gd32 clock config, may help for the generalization work. 
 
Some pattern for PLL clock IRC source:
  ```
  If IRC16M available:
      use IRC16M;
  If IRC8M available but IRC48M not available:
    use IRC8M
  If IRC8M and IRC48M both available:
    use IRC48M
 ```
| SoC | IRC8M | IRC16M | IRC48M |
| ---------- |  ---  |  ---  |  ---  |
|  GD32F4xx  |   -   |  Yes  |  Yes  |
|  GD32F403  |  Yes  |   -   |  Yes  |
|  GD32F3x0  |  Yes  |   -   |  Yes  |
|  GD32VF103 |  Yes  |   -   |   -   |
|  GD32E50x  |  Yes  |   -   |  Yes  |
|  GD32E23x  |  Yes  |   -   |   -   |
|  GD32E10x  |  Yes  |   -   |  Yes  |
|  GD32F1x0  |  Yes  |   -   |   -   |


> ``-`` means not available on this soc.

Note:
  1. GD32F3x0 and GD32E23x series have a dedicated ADC clock IRC28M.
  2. GD32F1x0 series have a dedicated ADC clock IRC14M (GD32F130 and GD32F150) or IRC28M (GD32F170 and GD32F190).

Refer:
 1. [GD32F4xx_User_Manual_Rev2.5](http://gd32mcu.com/data/documents/yingyongbiji/GD32F4xx_User_Manual_Rev2.5.pdf)
 2. [GD32F403_User_Manual_Rev2.5](http://gd32mcu.com/data/documents/yingyongbiji/GD32F403_User_Manual_Rev2.5.pdf)
 3. [GD32F3x0_User_Manual_Rev2.5](http://gd32mcu.com/data/documents/yingyongbiji/GD32F3x0_User_Manual_Rev2.5.pdf)
 4. [GD32VF103_User_Manual_Rev1.3](http://gd32mcu.com/data/documents/yingyongbiji/GD32VF103_User_Manual_Rev1.3.pdf)
 5. [GD32E50x_User_Manual_Rev1.4](http://gd32mcu.com/data/documents/yingyongbiji/GD32E50x_User_Manual_Rev1.4.pdf)
 6. [GD32E23x_User_Manual_Rev1.4](http://gd32mcu.com/data/documents/yingyongbiji/GD32E23x_User_Manual_Rev1.4.pdf)
 7. [GD32E10x_User_Manual_Rev1.6](http://gd32mcu.com/data/documents/yingyongbiji/GD32E10x_User_Manual_Rev1.6.pdf)
 8. [GD32F1x0_User_Manual_Rev3.5](http://gd32mcu.com/data/documents/yingyongbiji/GD32F1x0_User_Manual_Rev3.5.pdf)

File: drivers/clock_control/clock_control_gd32.c [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+{
+	uint16_t cfg = *(uint16_t *)sys;
+
+	switch (GD32_RCU_CFG_OFFSET(cfg)) {
+	case GD32_AHB1EN_OFFSET:
+		__fallthrough;
+	case GD32_AHB2EN_OFFSET:
+		__fallthrough;
+	case GD32_AHB3EN_OFFSET:
+		*rate = SystemCoreClock;
```
Comment (cameled) [inline drivers/clock_control/clock_control_gd32.c:228]: Looks ``break`` missing in here.

File: include/dt-bindings/clock/gd32_rcu_f4xx.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022 Teslabs Engineering S.L.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_GD32_RCU_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_GD32_RCU_H_
+
```
Comment (soburi) [inline include/dt-bindings/clock/gd32_rcu_f4xx.h:9]: ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_GD32_RCU**_F4XX**_H_

File: dts/bindings/clock/gd,gd32-rcu.yaml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        - 1
+        - 2
+        - 4
+        - 8
+        - 16
+        - 64
+        - 128
+        - 256
+        - 512
+      description: APB1 prescaler (sets CH_AHB frequency)
```
Comment (cameled) [inline dts/bindings/clock/gd,gd32-rcu.yaml:70]: AHB prescaler

**Comment by cameled** [comment]
@gmarull Any update? If you don't have much time to push this, maybe I can help.

**Pull request closed without merging**